### ISSUE TITLE
some online help/docu would be nice

### ISSUE THREAD

**Issue by dneise** [cta-sst-1m/histogram#1]
I was recently traveling on the train and had not internet.

### Preamble: 

I have this package installed as a dependency of digicampipe. This means it was basically installed like this on my machine:

    pip install https://github.com/calispac/histogram/archive/0.1.1.tar.gz

The result is that the c-sources are not copied to my machine, I only get compiled binary and the py-sources:
![2018-06-07-085320_1920x1080_scrot](https://user-images.githubusercontent.com/8200858/41083023-7090a278-6a30-11e8-9ef6-1e2febdd6e23.png)

So I cannot look into the c-sources when I do not understand something.

----

### Problem:

Well ... I simply did no know how to use this package. The problem has several phases. 

## Phase 1: what's inside again?

Basically all I remembered was the name of the package (something with `hist..`), so I imported it and wanted to see what was in it again ... I did not remember the class names for example, even though I was pretty sure there is only one class in this package: So I did this:

```python
# I actually typed hist<tab tab> and it was completed to histogram!
# So that worked! Very nice!
In [1]: import histogram  

In [2]: histogram.  # pressing <tab> .. multiple times ... nothing.

# trying something else:
In [2]: histogram?
Type:        module
String form: <module 'histogram' from '/home/dneise/anaconda3-5.1.0/envs/digicampipe/lib/python3.6/site-packages/histogram/__init__.py'>
File:        ~/anaconda3-5.1.0/envs/digicampipe/lib/python3.6/site-packages/histogram/__init__.py
Docstring:   <no docstring>

# So this does not help me finding how the classes are called in this package... again.
# Starting to feel a little helpless, so I try
In [3]: help(histogram)  
# Without much hope, since I think the `?` magic and `help()` do the same ... but no ... actually I get this:
```
![2018-06-07-090404_1920x1080_scrot](https://user-images.githubusercontent.com/8200858/41083574-cb5c4ee0-6a31-11e8-8fca-6949acf75a4e.png)

Ah ... I kind of remember now ... I have to do:

```python
In [7]: from histogram import histogram
# I actually typed: "histogram.<tab>" and got a list of things in this module
# So that worked very nice!
In [8]: histogram.Histogram1D?
Init signature: histogram.Histogram1D(bin_edges, data_shape=(1,), name='1D Histogram', axis_name=None)
Docstring:      <no docstring>
File:           ~/anaconda3-5.1.0/envs/digicampipe/lib/python3.6/site-packages/histogram/histogram.py
Type:           type
```

So this took me quite a while and at this point I am already a bit in a hurry, since all I wanted to do, is make a histogram and I remembered you wrote a nice package for this... and it is always better to reuse instead of writing it again.

----

## Phase 2: How did `Histogram1D` work again?

So now since I found the class name, I am pretty confident, that I can use it ... but I must admit, I do not really know what the init parameter `data_shape` is supposed to mean. Luckily it has a default argument so I ignore it for the moment.

Let's try to histogram a few random numbers from a normal distribution, just to see how the class works again ... 

```python
In [1]: from histogram.histogram import Histogram1D

In [2]: import numpy as np

In [3]: H = Histogram1D(bin_edges=np.linspace(-10, 10, 101))

In [4]: random_numbers = np.random.normal(loc=2, scale=1.5, size=1000)

In [5]: H.fill(random_numbers)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-5-9fe0f7c356f6> in <module>()
----> 1 H.fill(random_numbers)

~/anaconda3-5.1.0/envs/digicampipe/lib/python3.6/site-packages/histogram/histogram.py in fill(self, data_points)
     61         if data_points.shape[:-1] != self.shape[:-1]:
     62 
---> 63             raise IndexError
     64 
     65         new_first_axis = 1

IndexError: 

In [6]: H.fill?
Signature: H.fill(data_points)
Docstring:
:param data_points: ndarray, nan values are ignored
:return:
File:      ~/anaconda3-5.1.0/envs/digicampipe/lib/python3.6/site-packages/histogram/histogram.py
Type:      method

```

Hm ... at this point I am a bit unsure what is going on .. I have an array of numbers I want to histogram, but I get an IndexError. 

At this point I opened my editor ... and I looked through the code of the package .. but in one place there is a call into the C-part of this package and I absolutely do not remember what the C-part was doing...

So after quite some time I realize I am unable to use this class, give up and write it from scratch.

I assume some kind of documentation might have helped.

**Comment by calispac** [comment]
`data_shape` refers to the shape of the "arrays" of histograms. Let's say of have an experiment which has n_channels and then you take some data which will be an numpy array of shape `(n_channels, whatever_number)` and you a stream of these data. Let's suppose that the values are stored to a computer by a 16bits ADC. In that case you will have something like this: 

```python
import numpy as np
from histogram.histogram import Histogram1D

bin_edges = np.arange(0, 2**16, 1)
n_channels = 2048
data_shape = (n_channels, )
my_hist = Histogram1D(bin_edges=bin_edges, data_shape=data_shape)
for data in my_data_stream: 
    my_hist.fill(data)
```

Then you can acces the bin counts with:

```python
my_hist.data
```

Your index error comes from the fact that the input data to `fill()` has to have the shape `(1, whatever_number)` in your case it is `(whatever_number)`. This is probably something that I should fix at some point...

**Comment by calispac** [comment]
> So I cannot look into the c-sources when I do not understand something.

Well I think you would have it if you first did `git clone ...`. I suppose this is a choice of installation procedure. Whenever you want to run some code either you take the source and compile everything or you just take what is necessary. I personally prefer that `pip` does not copy all C source code to my `site-packages/`
